nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java

```diff
@@ -168,35 +231,54 @@ public boolean includeApInfoInSavedContext(AccessPath accessPath, VisitorState s
     return false;
   }
 
-  private void updateIfAssertIsPresentTrueOnOptional(
-      Context context,
-      MethodInvocationNode node,
-      Types types,
-      AccessPath.AccessPathContext apContext,
-      AccessPathNullnessPropagation.Updates bothUpdates) {
+  private Optional<MethodInvocationNode> getMethodWrappedByAssertThat(MethodInvocationNode node) {
     Node receiver = node.getTarget().getReceiver();
     if (receiver instanceof MethodInvocationNode) {
       MethodInvocationNode receiverMethod = (MethodInvocationNode) receiver;
       Symbol.MethodSymbol receiverSymbol = ASTHelpers.getSymbol(receiverMethod.getTree());
-      if (methodNameUtil.isMethodAssertThat(receiverSymbol)) {
-        // assertThat will always have at least one argument, So safe to extract from the arguments
+      if (receiverMethod.getArguments().size() == 1
+          && methodNameUtil.isMethodAssertThat(receiverSymbol)) {
         Node arg = receiverMethod.getArgument(0);
         if (arg instanceof MethodInvocationNode) {
-          // Since assertThat(a.isPresent()) changes to
-          // Truth.assertThat(Boolean.valueOf(a.isPresent()))
-          // need to be unwrapped from Boolean.valueOf
-          Node unwrappedArg = ((MethodInvocationNode) arg).getArgument(0);
-          if (unwrappedArg instanceof MethodInvocationNode) {
-            MethodInvocationNode argMethod = (MethodInvocationNode) unwrappedArg;
-            Symbol.MethodSymbol argSymbol = ASTHelpers.getSymbol(argMethod.getTree());
-            if (optionalIsPresentCall(argSymbol, types)) {
-              updateNonNullAPsForOptionalContent(
-                  context, bothUpdates, argMethod.getTarget().getReceiver(), apContext);
-            }
-          }
+          MethodInvocationNode wrappedMethodInvocation = ((MethodInvocationNode) arg);
+          return Optional.of(maybeUnwrapBooleanValueOf(wrappedMethodInvocation));
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  private Optional<Node> getVariableOrFieldWrappedByAssertThat(MethodInvocationNode node) {
+    Node receiver = node.getTarget().getReceiver();
+    if (receiver instanceof MethodInvocationNode) {
+      MethodInvocationNode receiverMethod = (MethodInvocationNode) receiver;
+      Symbol.MethodSymbol receiverSymbol = ASTHelpers.getSymbol(receiverMethod.getTree());
+      if (receiverMethod.getArguments().size() == 1
+          && methodNameUtil.isMethodAssertThat(receiverSymbol)) {
+        // assertThat will always have at least one argument, So safe to extract from the arguments
+        Node arg = receiverMethod.getArgument(0);
+        if (arg instanceof LocalVariableNode || arg instanceof FieldAccessNode) {
```

Comment:
are there any other valid cases we want find here?

i.e. `assertThat(obj.getOptionalField()).isPresent()` ?

Reply:
makes sense, can you provide guidance or existing code that shows how "access path" handling logic can be reused here?
it wasnt immediately clear to me at the `Node instanceof` level :sweat_smile:

Reply:
> (note that the key needs to be a constant for NullAway to understand that multiple calls to .get(...) will retrieve the same value)

wondering: why is that? is that a planned feature to at least support `final String key = "foo"; map.get(key)` or the like?
because that didnt seem to work when i tried writing a test.

Reply:
> > (note that the key needs to be a constant for NullAway to understand that multiple calls to .get(...) will retrieve the same value)
> 
> wondering: why is that? is that a planned feature to at least support `final String key = "foo"; map.get(key)` or the like? because that didnt seem to work when i tried writing a test.

This is not planned currently, but it's mostly a matter of priorities.  If that pattern comes up a lot in your code base (a `final` local variable assigned a string constant), I think we'd be open to a PR.  We'd just have to be careful it doesn't add significantly to NullAway's overhead.

Reply:
Ah, my bad here! We have such a support for `final static` fields and I think an open issue on supporting `final` local vars (see discussion in #697), I just got ahead of myself here 😅

Reply:
thanks for the feedback

>  If that pattern comes up a lot in your code base (a final local variable assigned a string constant), I think we'd be open to a PR

it imo doesnt have to be a string constant, just a type known to be immutable (that cant be re-assigned due to the final keyword).

here is an example of what i was unable to do in the test:
```
final String theKey = "theKey";
Assertions.assertThat(getMapWithOptional().get(theKey)).isNotNull();
Assertions.assertThat(getMapWithOptional().get(theKey)).isPresent();
int y = 1 + getMapWithOptional().get(theKey).get();
```
resulting in:
```
/Test.java:66: warning: [NullAway] Invoking get() on possibly empty Optional getMapWithOptional().get(theKey)
    int y = 1 + getMapWithOptional().get(theKey).get();
                                                ^
```
but it would work when using the same string literal everywhere as key instead.
is the access path not considered equal when a variable is involved like that?
that's what i was wondering about here.

i dont immediately see how this relates to the Stream-issue which adds another layer of complexity.

Reply:
I agree this is unrelated to streams. I don't immediately see any conceptual issues with treating final local variables as constant keys in map access paths. @lazaroclapp do you see an issue? Not too sure about complexity of implementing support or perf cost

Reply:
Ok, on the linked issue, the relevant comment is [this one](https://github.com/uber/NullAway/issues/697#issuecomment-1350129124). But I entirely agree that the context of that issue overcomplicates the discussion. So here is my best attempt of an explanation:

1. During intra-procedural dataflow, there is a set of locations NullAway tracks the nullability of, defined by the concept of an Access Path. This includes: variables (locals, arguments, etc), chains of fields (`this.f.g.h`), chains of method calls that NullAway interprets to be getters through some heuristics (the easies case is zero-argument methods: `this.getF().getG().getH()`), and - as a semi-special case - `Map::get(k)` for a constant key `k`. 
2. Combinations of the above are allowed. A concatenation of an existing valid Access Path followed by one of the above is also a valid Access Path. Here is the [data structure](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java) used, which gives some context about the cases. 
3. What "a constant key `k`" means, though, is maybe a bit hacky. Primitive constants are acknowledged as constant here, as are `static fields` fields. Here is an example [test](https://github.com/uber/NullAway/blob/master/nullaway/src/test/java/com/uber/nullaway/NullAwayAccessPathsTests.java#L81) of what is known to be supported.
4. Now, interesting caveat from your example above. I see that `Assertions.assertThat(getMapWithOptional().get(theKey)).isNotNull()` is being interpreted correctly as guaranteeing the path `getMapWithOptional().get(theKey)` is non-null, or else you should be getting a standard nullness error for that final line, rather than the empty optional issue. No? Or is it just that the optional emptiness error triggers first somehow? If the former, then that is a bit surprising to me, based on what we test for, but to be fair there does seem to be a [more general case](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java#L297) for map keys that would match such expression. 

What I would do, if I were debugging this at this point, would be to check:

1. If the example above, with a `final String thisKey` local variable, works when only checking for the nullability of the map's value, independent of optional emptiness handling.
2. If the example above, with optional emptiness handling on, works when `theKey` is a `static final` field, as in the test linked above.

Either way, I would leave any fix for a follow up PR, because if this ends up involving changing core Access Path definition stuff rather than just this handler, then we need to be very careful about performance/end-to-end testing and it's easier with a smaller change. And, even if it doesn't, it feels like conceptually separate from supporting more test assertion APIs 🙂